File: tools/hangar/src-tauri/src/fleetctl.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::process::Stdio;
use tokio::io::AsyncWriteExt;
use tokio::process::Command;

fn shellexpand(s: &str) -> String {
    if let Some(rest) = s.strip_prefix("~/") {
        if let Some(home) = dirs::home_dir() {
            return home.join(rest).to_string_lossy().to_string();
        }
    }
    if s == "~" {
        if let Some(home) = dirs::home_dir() {
            return home.to_string_lossy().to_string();
        }
    }
    s.to_string()
}

#[derive(Debug, Serialize)]
pub struct ResolvedBinary {
    pub path: String,
    pub source: &'static str, // "settings" | "build" | "missing"
    pub exists: bool,
}

#[tauri::command(rename_all = "snake_case")]
pub fn fleetctl_resolve_binary(
    repo: Option<String>,
    settings_path: Option<String>,
) -> Result<ResolvedBinary, String> {
    // Prefer the explicit settings path if the user picked one — they
    // may be testing a release binary outside the repo. Fall back to
    // the conventional <repo>/build/fleetctl so the common dev path
    // just works.
    if let Some(p) = settings_path.as_deref().filter(|s| !s.is_empty()) {
        let expanded = shellexpand(p);
        let exists = Path::new(&expanded).exists();
        return Ok(ResolvedBinary {
            path: expanded,
            source: "settings",
            exists,
        });
    }
    if let Some(r) = repo.as_deref().filter(|s| !s.is_empty()) {
        let candidate = Path::new(r).join("build").join("fleetctl");
        let exists = candidate.exists();
        return Ok(ResolvedBinary {
            path: candidate.to_string_lossy().to_string(),
            source: "build",
            exists,
        });
    }
    Ok(ResolvedBinary {
        path: String::new(),
        source: "missing",
        exists: false,
    })
}

// ----- ~/.fleet/config parsing -----

#[derive(Debug, Deserialize)]
struct FleetConfigFile {
    // serde_yaml::Mapping preserves insertion order (it's backed by
    // indexmap internally). We use it instead of HashMap so the list
    // we return mirrors the order of contexts in the YAML file —
    // which is what the user sees in the editor and expects to match
    // the parsed list panel.
    #[serde(default)]
    contexts: serde_yaml::Mapping,
}

#[derive(Debug, Deserialize, Default)]
struct RawContext {
    #[serde(default)]
    address: Option<String>,
    #[serde(default)]
    email: Option<String>,
    #[serde(default)]
    token: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct ContextSummary {
    pub name: String,
    pub address: Option<String>,
    pub email: Option<String>,
    pub has_token: bool,
}

#[derive(Debug, Serialize)]
pub struct ContextInfo {
    pub config_path: String,
    pub exists: bool,
    pub current: Option<ContextSummary>,
    pub contexts: Vec<ContextSummary>,
}

fn default_config_path() -> PathBuf {
    if let Some(home) = dirs::home_dir() {
        home.join(".fleet").join("config")
    } else {
        PathBuf::from("~/.fleet/config")
    }
}

#[derive(Debug, Serialize)]
pub struct RawConfig {
    pub path: String,
    pub exists: bool,
    pub contents: String,
}

#[tauri::command]
pub fn fleetctl_read_config_raw() -> Result<RawConfig, String> {
    let path = default_config_path();
    let path_str = path.to_string_lossy().to_string();
    if !path.exists() {
        return Ok(RawConfig {
            path: path_str,
            exists: false,
            contents: String::new(),
        });
    }
    let raw = std::fs::read_to_string(&path).map_err(|e| e.to_string())?;
    Ok(RawConfig {
        path: path_str,
        exists: true,
        contents: raw,
    })
}

#[tauri::command]
pub fn fleetctl_save_config(yaml: String) -> Result<(), String> {
    // Validate parse first so users see "line 12: unexpected mapping"
    // rather than fleetctl crashing later. We just need the file to be
    // valid YAML — don't enforce structure beyond that; the user may
    // be doing something unusual with custom-headers or similar.
    serde_yaml::from_str::<serde_yaml::Value>(&yaml)
        .map_err(|e| format!("YAML parse error: {e}"))?;
    let path = default_config_path();
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)
            .map_err(|e| format!("creating {parent:?}: {e}"))?;
    }
    std::fs::write(&path, yaml).map_err(|e| format!("writing {path:?}: {e}"))
}

#[tauri::command]
pub fn fleetctl_read_context() -> Result<ContextInfo, String> {
    // We always read the *default* config path because fleetctl itself
    // does the same unless --config is passed. If the user runs the
    // binary with a custom CONFIG env elsewhere, this won't reflect
    // it — but Fleet Hangar always launches fleetctl with no overrides,
    // so the two stay aligned.
    let path = default_config_path();
    let path_str = path.to_string_lossy().to_string();
    if !path.exists() {
        return Ok(ContextInfo {
            config_path: path_str,
            exists: false,
            current: None,
            contexts: Vec::new(),
        });
    }
    let raw = std::fs::read_to_string(&path).map_err(|e| e.to_string())?;
    let parsed: FleetConfigFile = serde_yaml::from_str(&raw)
        .map_err(|e| format!("parse {path_str}: {e}"))?;
    // Walk the Mapping in iteration order — which is file order. We
    // skip entries whose key isn't a string (malformed config) and
    // skip values that can't be deserialized into a RawContext rather
    // than failing the whole read.
    let contexts: Vec<ContextSummary> = parsed
        .contexts
        .into_iter()
        .filter_map(|(k, v)| {
            let name = k.as_str()?.to_string();
            let raw: RawContext = serde_yaml::from_value(v).unwrap_or_default();
            Some(ContextSummary {
                name,
                address: raw.address,
                email: raw.email,
                has_token: raw
                    .token
                    .as_deref()
                    .map(|t| !t.is_empty())
                    .unwrap_or(false),
            })
        })
        .collect();
    // fleetctl's default context name is "default" unless the user
    // selected something else with `fleetctl config switch-context` —
    // which writes to a different file we don't currently track. For
    // Fleet Hangar dev, "default" is right.
    let current = contexts.iter().find(|c| c.name == "default").cloned();
    Ok(ContextInfo {
        config_path: path_str,
        exists: true,
        current,
        contexts,
    })
}

impl Clone for ContextSummary {
    fn clone(&self) -> Self {
        Self {
            name: self.name.clone(),
            address: self.address.clone(),
            email: self.email.clone(),
            has_token: self.has_token,
        }
    }
}

// ----- one-shot capture runner -----

#[derive(Debug, Serialize)]
pub struct CapturedRun {
    pub exit_code: Option<i32>,
    pub stdout: String,
    pub stderr: String,
}

/// Runs a fleetctl invocation synchronously and returns the captured
/// output. Use this for short, finite commands where the caller wants
/// the result inline (login, get, status checks). For long-running or
/// streamy commands, use the existing start_process pipeline instead so
/// output shows up in the Logs tab.
#[tauri::command(rename_all = "snake_case")]
pub async fn fleetctl_run_capture(
    program: String,
    cwd: Option<String>,
    args: Vec<String>,
    env: Option<HashMap<String, String>>,
    stdin_data: Option<String>,
    timeout_ms: Option<u64>,
) -> Result<CapturedRun, String> {
    let mut cmd = Command::new(&program);
    cmd.args(&args)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .stdin(if stdin_data.is_some() {
            Stdio::piped()
        } else {
            Stdio::null()
        });
    // Login-shell PATH so fleetctl (and anything it shells out to)
    // resolves when launched from Finder. Set before caller env so an
    // explicit PATH from the caller would still take precedence.
    cmd.env("PATH", crate::shellpath::shell_path());
    if let Some(dir) = cwd.as_deref().filter(|s| !s.is_empty()) {
        cmd.current_dir(dir);
    }
    if let Some(envs) = env {
        for (k, v) in envs {
            cmd.env(k, v);
        }
    }

    let mut child = cmd
        .spawn()
        .map_err(|e| format!("spawn {program}: {e}"))?;

    if let Some(data) = stdin_data {
        if let Some(mut stdin) = child.stdin.take() {
            let _ = stdin.write_all(data.as_bytes()).await;
            let _ = stdin.shutdown().await;
        }
    }

    let timeout = std::time::Duration::from_millis(timeout_ms.unwrap_or(60_000));
    let waited = tokio::time::timeout(timeout, child.wait_with_output()).await;
    let out = match waited {
        Ok(Ok(o)) => o,
        Ok(Err(e)) => return Err(format!("wait {program}: {e}")),
        Err(_) => {
            // Best-effort cleanup; the kill_on_drop wasn't set here.
            return Err(format!("timed out after {}ms", timeout.as_millis()));
        }
    };
    Ok(CapturedRun {
        exit_code: out.status.code(),
        stdout: String::from_utf8_lossy(&out.stdout).to_string(),
        stderr: String::from_utf8_lossy(&out.stderr).to_string(),
    })
}

File: tools/hangar/internal/fleetctl/fleetctl.go
// Package fleetctl resolves the fleetctl binary, reads/writes the fleetctl
// config (~/.fleet/config), and runs one-shot fleetctl invocations with
// captured output. Ported from src-tauri/src/fleetctl.rs.
package fleetctl

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/fleetdm/fleet/tools/hangar/internal/paths"
	"github.com/fleetdm/fleet/tools/hangar/internal/shellpath"
	"gopkg.in/yaml.v3"
)

// ResolvedBinary reports where fleetctl was found and how.
type ResolvedBinary struct {
	Path   string `json:"path"`
	Source string `json:"source"` // "settings" | "build" | "missing"
	Exists bool   `json:"exists"`
}

// ResolveBinary prefers an explicit settings path (the user may be testing a
// release binary), then <repo>/build/fleetctl, then reports missing.
func ResolveBinary(repo, settingsPath string) ResolvedBinary {
	if settingsPath != "" {
		expanded := paths.Expand(settingsPath)
		return ResolvedBinary{Path: expanded, Source: "settings", Exists: fileExists(expanded)}
	}
	if repo != "" {
		candidate := filepath.Join(repo, "build", "fleetctl")
		return ResolvedBinary{Path: candidate, Source: "build", Exists: fileExists(candidate)}
	}
	return ResolvedBinary{Source: "missing"}
}

func fileExists(p string) bool {
	_, err := os.Stat(p)
	return err == nil
}

// ContextSummary is one fleetctl context (sensitive token reduced to a bool).
type ContextSummary struct {
	Name     string  `json:"name"`
	Address  *string `json:"address"`
	Email    *string `json:"email"`
	HasToken bool    `json:"has_token"`
}

// ContextInfo describes the parsed fleetctl config.
type ContextInfo struct {
	ConfigPath string           `json:"config_path"`
	Exists     bool             `json:"exists"`
	Current    *ContextSummary  `json:"current"`
	Contexts   []ContextSummary `json:"contexts"`
}

// RawConfig is the raw fleetctl config file contents.
type RawConfig struct {
	Path     string `json:"path"`
	Exists   bool   `json:"exists"`
	Contents string `json:"contents"`
}

// DefaultConfigPath is ~/.fleet/config — what fleetctl reads with no --config.
func DefaultConfigPath() string {
	home, err := os.UserHomeDir()
	if err != nil {
		return "~/.fleet/config"
	}
	return filepath.Join(home, ".fleet", "config")
}

type rawContext struct {
	Address *string `yaml:"address"`
	Email   *string `yaml:"email"`
	Token   *string `yaml:"token"`
}

// parseContexts walks the contexts mapping in file order (using a yaml.Node
// so we don't lose ordering the way a Go map would), skipping non-string
// keys and undecodable values rather than failing the whole read.
func parseContexts(raw []byte) ([]ContextSummary, error) {
	var doc struct {
		Contexts yaml.Node `yaml:"contexts"`
	}
	if err := yaml.Unmarshal(raw, &doc); err != nil {
		return nil, err
	}
	out := []ContextSummary{}
	c := doc.Contexts
	if c.Kind != yaml.MappingNode {
		return out, nil
	}
	for i := 0; i+1 < len(c.Content); i += 2 {
		keyNode, valNode := c.Content[i], c.Content[i+1]
		if keyNode.Kind != yaml.ScalarNode {
			continue
		}
		var rc rawContext
		_ = valNode.Decode(&rc) // tolerate malformed entries (unwrap_or_default)
		out = append(out, ContextSummary{
			Name:     keyNode.Value,
			Address:  rc.Address,
			Email:    rc.Email,
			HasToken: rc.Token != nil && *rc.Token != "",
		})
	}
	return out, nil
}

// ReadConfigRaw returns the raw fleetctl config contents (empty if absent).
func ReadConfigRaw(configPath string) (RawConfig, error) {
	b, err := os.ReadFile(configPath)
	if os.IsNotExist(err) {
		return RawConfig{Path: configPath, Exists: false}, nil
	}
	if err != nil {
		return RawConfig{}, err
	}
	return RawConfig{Path: configPath, Exists: true, Contents: string(b)}, nil
}

// SaveConfig validates that yaml parses, then writes it to configPath
// (creating the parent dir).
func SaveConfig(configPath, yamlText string) error {
	var probe any
	if err := yaml.Unmarshal([]byte(yamlText), &probe); err != nil {
		return fmt.Errorf("YAML parse error: %w", err)
	}
	if parent := filepath.Dir(configPath); parent != "" {
		if err := os.MkdirAll(parent, 0o755); err != nil {
			return fmt.Errorf("creating %s: %w", parent, err)
		}
	}
	return os.WriteFile(configPath, []byte(yamlText), 0o644)
}

// ReadContext parses the fleetctl config at configPath. The current context
// is "default" (Hangar always launches fleetctl with no overrides).
func ReadContext(configPath string) (ContextInfo, error) {
	b, err := os.ReadFile(configPath)
	if os.IsNotExist(err) {
		return ContextInfo{ConfigPath: configPath, Exists: false, Contexts: []ContextSummary{}}, nil
	}
	if err != nil {
		return ContextInfo{}, err
	}
	contexts, err := parseContexts(b)
	if err != nil {
		return ContextInfo{}, fmt.Errorf("parse %s: %w", configPath, err)
	}
	var current *ContextSummary
	for i := range contexts {
		if contexts[i].Name == "default" {
			c := contexts[i]
			current = &c
			break
		}
	}
	return ContextInfo{ConfigPath: configPath, Exists: true, Current: current, Contexts: contexts}, nil
}

// CapturedRun is the result of a one-shot command.
type CapturedRun struct {
	ExitCode *int   `json:"exit_code"`
	Stdout   string `json:"stdout"`
	Stderr   string `json:"stderr"`
}

// RunCapture runs program synchronously with the login-shell PATH (plus any
// caller env), optional cwd and stdin, and a timeout (default 60s). Use for
// short, finite commands; long/streamy commands go through the process
// engine so output lands in the Logs tab.
func RunCapture(program, cwd string, args []string, env map[string]string, stdinData string, timeoutMS uint64) (CapturedRun, error) {
	if timeoutMS == 0 {
		timeoutMS = 60_000
	}
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutMS)*time.Millisecond)
	defer cancel()

	cmd := shellpath.CommandContext(ctx, program, args...)
	cmd.Env = shellpath.MergeEnv(cmd.Env, env)
	if cwd != "" {
		cmd.Dir = cwd
	}
	if stdinData != "" {
		cmd.Stdin = strings.NewReader(stdinData)
	}
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	if ctx.Err() == context.DeadlineExceeded {
		return CapturedRun{}, fmt.Errorf("timed out after %dms", timeoutMS)
	}

	run := CapturedRun{Stdout: stdout.String(), Stderr: stderr.String()}
	if err == nil {
		zero := 0
		run.ExitCode = &zero
		return run, nil
	}
	var ee *exec.ExitError
	if errors.As(err, &ee) {
		// Exited normally → real code; killed by signal → leave nil (matches
		// Rust's ExitStatus::code() returning None on signal termination).
		if ee.ProcessState.Exited() {
			code := ee.ExitCode()
			run.ExitCode = &code
		}
		return run, nil
	}
	return CapturedRun{}, fmt.Errorf("spawn %s: %w", program, err)
}
